### PR TITLE
CLOUDSTACK-9708: Router deployment failed due to two threads start VR simultaneously.

### PR DESCRIPTION
**Pull request by nitin-maharana**
Description:
=========
Please see the description in https://issues.apache.org/jira/browse/CLOUDSTACK-9708

Resolution:
========
- The  SearchBuilder instance is a class scope variable that should be instantiated in the constructor.
- Since it was being instantiated in the scope of the method, parallel calls to the method will overwrite the class variable.

### CODE REVIEW

**Comment by remibergsma** [comment]
@nitin-maharana Can you please double check this PR? Lots of tests are failing.

**Comment by nitin-maharana** [comment]
@remibergsma : Can you please review it again. Fixed the test failures.

ping @jayapalu @rajesh-battala

**Comment by jayapalu** [comment]
LGTM

**Comment by koteswararavipati** [comment]
LGTM

**Comment by nitin-maharana** [comment]
Thanks @koteswararavipati @jayapalu.

tag:mergeready

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 686
 Hypervisor xenserver
 NetworkType Advanced
 Passed=53
 Failed=4
 Skipped=11

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_network.py

 * test_network_rules_acquired_public_ip_1_static_nat_rule Failed

 * test_network_rules_acquired_public_ip_2_nat_rule Failed

 * test_network_rules_acquired_public_ip_3_Load_Balancer_Rule Failed

 * ContextSuite context=TestRouterRules>:teardown Failed


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_affinity_groups_projects.py
test_vm_snapshots.py
test_over_provisioning.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_non_contigiousvlan.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_network_acl.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 691
 Hypervisor xenserver
 NetworkType Advanced
 Passed=111
 Failed=1
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_ssvm.py

 * test_03_ssvm_internals Failed


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_routers_network_ops.py
test_disk_offerings.py

**Comment by jayapalu** [comment]
In the tests telnet to host port 8250 is failed. I think this can be ignore because in the  setup templates got downloaded and vm got deployed that means the ssvm agent is up. May be during the test there is agent connection fluctuation. It looks intermittent issue. 
Tagging merge ready. 


 /usr/local/cloud/systemvm/ssvm-check.sh |grep -e ERROR -e WARNING -e FAIL via Host: 10.147.28.52} {returns: [u'ERROR: Cannot connect to 10.147.28.145 port 8250']}&#10;paramiko.transport: DEBUG: EOF in transport thread&#10;test_03_ssvm_internals (integration.smoke.test_ssvm.TestSSVMs): DEBUG: SSVM script output: [u'ERROR: Cannot connect to 10.147.28.145 port 8250']&#10;test_03_ssvm_internals (integration.smoke.test_ssvm.TestSSVMs): CRITICAL: FAILED: test_03_ssvm_internals:

File: engine/schema/src/com/cloud/vm/dao/UserVmDaoImpl.java [hunk 1]
```diff
@@ -192,6 +192,18 @@ void init() {
                 JoinBuilder.JoinType.INNER);
         AccountDataCenterVirtualSearch.done();
 
+        SearchBuilder<NicVO> nicSearchByNetwork = _nicDao.createSearchBuilder();
```
Comment (rhtyd) [inline engine/schema/src/com/cloud/vm/dao/UserVmDaoImpl.java]: @nitin-maharana while your fix may be valid, it seems to be breaking some test_nic and test_routers test cases. The issue seems to be another `nicSearch` on line 184 which perhaps is used in most cases. You may remove `nicSearchByNetwork` which is not useful or used at all depending on the original code when UserVmSearch is null or not, when calling the method `listByNetworkIdAndStates`.  Depending on the test results, I may revert your original fix, but kindly re-submit a newer change and I can help run tests for you. Thanks.

**Comment by rhtyd** [comment]
@nitin-maharana This is from a test env running tests with the revert on #2220. Based on the results, I'll proceed with reverting of the change and fwd merge the reverts on 4.9, 4.10, master branches. Kindly re-submit the fix consider my comment above. Thanks.

[root@pr2220-t1309-vmware-55u3-marvin marvin]# python tools/xunit-reader.py MarvinLogs/ 

Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_09_reboot_router | Success | 160.79 | test_routers.py
test_08_start_router | Success | 135.70 | test_routers.py
test_07_stop_router | Success | 25.17 | test_routers.py
test_06_router_advanced | Success | 0.04 | test_routers.py
test_05_router_basic | Success | 0.03 | test_routers.py
test_04_restart_network_wo_cleanup | Success | 5.54 | test_routers.py
test_03_restart_network_cleanup | Success | 180.90 | test_routers.py
test_02_router_internal_adv | Success | 0.85 | test_routers.py
test_01_router_internal_basic | Success | 0.42 | test_routers.py
test_router_dns_guestipquery | Success | 76.72 | test_router_dns.py
test_router_dns_externalipquery | Success | 0.06 | test_router_dns.py
test_router_dhcphosts | Success | 135.51 | test_router_dhcphosts.py
test_router_dhcp_opts | Success | 21.36 | test_router_dhcphosts.py
test_01_nic | Success | 534.47 | test_nic.py
test_reboot_router | Success | 618.97 | test_network.py
test_nic_secondaryip_add_remove | Success | 242.11 | test_multipleips_per_nic.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py

File: engine/schema/src/com/cloud/vm/dao/UserVmDaoImpl.java (hunk 1 above)
Comment (nitin-maharana) [inline engine/schema/src/com/cloud/vm/dao/UserVmDaoImpl.java]: @rhtyd: Perfect! Thanks for finding out. I will submit a new pull request with the proper fix.

**Pull request merged**